IN-553 speed up group membership queries

kogre opened the pull request:
(IN-553)

## What changes are in this PR?

This PR is about speeding up queries that calculate user permissions on projects (granular permissions and project access).

We have discovered an inefficiency in how the platform calculates which users have access to which projects. Instead of first checking which groups have access to a project, and then checking whether a user is part of one of these groups, we're doing it the other way around: We first calculate all the groups the user is part of, and then see whether any of these groups overlap with the groups that have access to the project.

For some platforms, they have a lot of smart groups. This means that for the database, it's a difficult calculation (especially the "participated in" kind of smart group rules). This PR tries to chance the logic to only check the groups used for permissions, not all smart groups.

kogre committed:
Allow limiting the smart groups to search for membership in

kogre committed:
Removed unused Group#member? method

kogre committed:
Added User#in_any_groups?(groups)

kogre committed:
Use in_any_groups? for project #show permissions

kogre committed:
Merge branch 'master' into IN-553-speed-up-group-membership-queries

kogre commented on back/app/models/group.rb:
```diff
@@ -20,10 +20,6 @@ class Group < ApplicationRecord
-  def member?(user)
-    users.exists?(id: user.id)
-  end
-
```
Seems to be unused, so removed it.

kogre commented on back/engines/commercial/smart_groups/app/services/smart_groups/rules_service.rb:
```diff
@@ -93,5 +85,17 @@ def filter_by_value_references value, groups=nil
+    def group_if_users_included(users, group)
+      ::Group.where(id: group.id)
+             .where(filter(users, group.rules).arel.exists)
+    end
```
Moved these down in the file to make them private methods

adessy commented:
Hey @kogre, is there anything left to optimize here, or is it just a matter of testing and wrapping things up?

Do we have an example query to check / measure the improvement?

kogre commented:
Yes, a lot to improve further.

So far it is only affecting the project_policy#show implementation.

A similar query (calling into `user.groups` or `user.group_ids`) is also happening in the project_policy scope and if I remember correctly, from the `Permission` model, or related to that. All of these are triggering the same issue.

There is no test to completely validate that this is going well, but if I had to describe such test: It needs to check that whenever we're calling user.groups or user.group_ids, we really need all the groups a user is in, and not just a subset.

adessy committed:
IN-553 Avoid computing all group memberships

IN-553 #time 0w 0d 2h 59m Avoid computing all group memberships

adessy committed:
IN-553 Cosmetic

IN-553 #time 0w 0d 0h 29m Cosmetic

adessy committed:
IN-553 Merge extensions/group.rb into patches/group.rb

IN-553 #time 0w 0d 0h 18m Merge extensions/group.rb into patches/group.rb

adessy committed:
IN-553 Cosmetic

IN-553 #time 0w 0d 0h 2m Cosmetic

adessy committed:
IN-553 Add the smart group equivalent to manual_groups method + memoize the rules service

IN-553 #time 0w 0d 0h 5m Add the smart group equivalent to manual_groups method + memoize the rules service

adessy committed:
IN-553 Clarifying return types

IN-553 #time 0w 0d 0h 10m Clarifying return types

adessy committed:
Merge remote-tracking branch 'origin/master' into IN-553-speed-up-group-membership-queries

IN-553 #time 0w 3d 5h 58m Merge remote-tracking branch 'origin/master' into IN-553-speed-up-group-membership-queries

adessy committed:
IN-553 Add `with_user` scope to Group

IN-553 #comment This scope function is somehow similar to `SmartGroups::RulesService.new.groups_for_user` but the latter only support smart groups. The new scope is meant to work with any type of group.

IN-553 #time 0w 0d 6h 35m Add `with_user` scope to Group

adessy committed:
IN-553 Rework `user_groups_visible` scope to avoid calling `user.group_ids` which would resolve all smart groups

IN-553 #comment Now, it should only resolve the groups relevant for the project scope at hand.

IN-553 #time 0w 0d 0h 7m Rework `user_groups_visible` scope to avoid calling `user.group_ids` which would resolve all smart groups

adessy committed:
IN-553 Rework `for_user` scope to avoid calling `user.group_ids` which would resolve all smart groups

IN-553 #comment Now, it should only resolve the groups relevant for the permission scope at hand.

IN-553 #time 0w 0d 1h 14m Rework `for_user` scope to avoid calling `user.group_ids` which would resolve all smart groups

adessy committed:
IN-553 Simplify ProjectPolicy patching with our current best practices

IN-553 #time 0w 0d 0h 5m Simplify ProjectPolicy patching with our current best practices

adessy committed:
IN-553 Rubocop

IN-553 #time 0w 0d 0h 5m Rubocop

adessy committed:
IN-553 Fix `PG::UndefinedColumn: ERROR:  column groups.permission does not exist`

IN-553 #time 0w 0d 0h 44m Fix `PG::UndefinedColumn: ERROR:  column groups.permission does not exist`

adessy committed:
IN-553 Force early loading of permissions

IN-553 #comment Otherwise the difference between old and new permissions is always empty.

IN-553 #time 0w 0d 1h 30m Force early loading of permissions

adessy committed:
IN-553 Improve test readability

IN-553 #time 0w 0d 0h 16m Improve test readability

adessy committed:
IN-553 Add comments

IN-553 #time 0w 0d 0h 6m Add comments

adessy committed:
IN-553 Replace unnecessary `includes` with `joins`

IN-553 #time 0w 0d 0h 8m Replace unnecessary `includes` with `joins`

adessy commented:
Hey @kogre 👋 Could you review this one please? (I cannot add you as a reviewer bc you are the PR author.) It was kinda a mess to solve to say the least and I'm not sure to which extent it solves the problem. Do you think there are other areas that could use some refactoring?

adessy committed:
Revert "IN-553 Replace unnecessary `includes` with `joins`"

IN-553 #comment This reverts commit b326c24204f857e56592d2ac307c81965888910e.

IN-553 #time 0w 0d 0h 15m Revert "IN-553 Replace unnecessary `includes` with `joins`"

adessy commented on back/app/controllers/web_api/v1/users_controller.rb:
```diff
@@ -116,7 +116,7 @@ def create
 
   def update
-    permissions_before = Permission.for_user(@user)
+    permissions_before = Permission.for_user(@user).load
```
We need to load the records early to be able to compare permissions before and after.

adessy commented on back/app/models/project.rb:
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Mostly rubocop in this file

adessy commented on back/app/models/project.rb:
```diff
@@ -151,15 +153,17 @@ class Project < ApplicationRecord
   }
 
   scope :user_groups_visible, lambda { |user|
-    where("projects.visible_to = 'groups' AND EXISTS(SELECT 1 FROM groups_projects WHERE project_id = projects.id AND group_id IN (?))", user.group_ids)
+    user_groups = Group.joins(:projects).where(projects: self).with_user(user)
+    project_ids = GroupsProject.where(projects: self).where(groups: user_groups).select(:project_id).distinct
+    where(id: project_ids)
```
to be reviewed

adessy commented on back/app/policies/project_policy.rb:
```diff
@@ -155,6 +157,5 @@ def moderate_for_active?
 ProjectPolicy.prepend(Polls::Patches::ProjectPolicy)
 
 ProjectPolicy.prepend_if_ee('ProjectFolders::Patches::ProjectPolicy')
-ProjectPolicy::Scope.prepend_if_ee('ProjectFolders::Patches::ProjectPolicy::Scope')
```
Applying our (my?) patching guidelines.

adessy commented on back/engines/commercial/project_folders/app/policies/project_folders/patches/project_policy.rb:
```diff
@@ -3,6 +3,10 @@
 module ProjectFolders
   module Patches
     module ProjectPolicy
+      def self.prepended(base)
+        base::Scope.prepend(Scope)
+      end
```
Applying our patching guidelines.

adessy commented on back/spec/acceptance/users_spec.rb:
```diff
@@ -570,34 +570,33 @@
       let(:first_name) { "Edmond" }
```
This test was failing because of the missing `load` in `users_controller.rb`. I just reworked it a bit along the way to improve readability.

kogre commented on back/app/models/permission.rb:
```diff
@@ -32,19 +32,21 @@ class Permission < ApplicationRecord
-    else
-      where(permitted_by: 'everyone')
-    end
+    next where(permitted_by: 'everyone') unless user
+    next all if user.admin?
```
I know it's a rubocop rule and 99% personal perference, but just to get it off my chest: I find this so much harder to read compared to the original if/else branching.

adessy replied:
Hehe, I see what you mean. Personally, I don't have a big preference for one or the other. In this specific case, it saves me an extra indentation and compensates a bit the lengthy lines of the scope. But I would definitely not use guards everywhere.

kogre reviewed (commented):
I can't approve because I own it, but it looks good to me. Database servers around the world will sigh in relief when the code lands!

I focused on the critical sections, and I think the logic is sound. How confident are you? Any tests you can think of that could boost it further? (except for some manual tests, we should definitely do that in this case, best to make an epic and ask a review!)

My only point of doubt is whether we aren't introducing any N+1 query pattern now, if some of this code is called in a loop. Because of some `.ids` calls we risk firing off the query. But that should be solved with profiling on critical endpoints, as it's almost impossible to deduct it.

adessy commented:
> Database servers around the world will sigh in relief when the code lands!

❤️‍🔥  

> How confident are you? 

Not as confident as usual, but still reasonably confident.

> Any tests you can think of that could boost it further? (except for some manual tests, we should definitely do that in this case, best to make an epic and ask a review!)

That's a refactoring, so I assumed it already enjoyed some test coverage 😇 Actually, my first attempts broke some of them. So the assumption was not too far fetched 😄 I can setup an epic platform. Should I ask Karel for a review?

> My only point of doubt is whether we aren't introducing any N+1 query pattern now, if some of this code is called in a loop. Because of some `.ids` calls we risk firing off the query. But that should be solved with profiling on critical endpoints, as it's almost impossible to deduct it.

Not sure to follow you here. Do you have a minimalistic example in mind where this would happen?

adessy committed:
IN-553 Rubocop

IN-553 #time 0w 3d 1h 7m Rubocop
